src/Templates/BlockTemplatesCompatibility.php

```diff
@@ -0,0 +1,290 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Templates;
+
+/**
+ * BlockTemplatesCompatibility class.
+ *
+ * To bridge the gap on compatibility with PHP hooks and blockified templates.
+ *
+ * @internal
+ */
+class BlockTemplatesCompatibility {
+
+	const LOOP_ITEM_ID = 'product-loop-item';
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->init();
+	}
+
+	/**
+	 * Initialization method.
+	 */
+	protected function init() {
+		if ( ! wc_current_theme_is_fse_theme() ) {
+			return;
+		}
+
+		add_filter( 'render_block_data', array( $this, 'update_render_block_data' ), 10, 3 );
+		add_filter( 'render_block', array( $this, 'inject_hooks' ), 10, 2 );
+	}
+
+	/**
+	 * Update the render block data to inject our custom attribute needed to
+	 * determine which blocks belong to an inherited Products block.
+	 *
+	 * @param array         $parsed_block The block being rendered.
+	 * @param array         $source_block An un-modified copy of $parsed_block, as it appeared in the source content.
+	 * @param WP_Block|null $parent_block If this is a nested block, a reference to the parent block.
+	 *
+	 * @return array
+	 */
+	public function update_render_block_data( $parsed_block, $source_block, $parent_block ) {
+
+		if ( ! $this->is_archive_template() ) {
+			return $parsed_block;
+		}
+
+		/**
+		 * Custom data can be injected to top level block only, as Gutenberg
+		 * will use this data to render the blocks and its nested blocks.
+		 */
+		if ( $parent_block ) {
+			return $parsed_block;
+		}
+
+		array_walk( $parsed_block['innerBlocks'], array( $this, 'inner_blocks_walker' ) );
+
+		return $parsed_block;
+	}
+
+	/**
+	 * Inject hooks to rendered content of corresponding blocks.
+	 *
+	 * @param mixed $block_content The rendered block content.
+	 * @param mixed $block         The parsed block data.
+	 * @return string
+	 */
+	public function inject_hooks( $block_content, $block ) {
+		if ( ! $this->is_archive_template() ) {
+			return $block_content;
+		}
+		/**
+		 * If the block is not inherited, we don't need to inject hooks.
+		 */
+		if ( empty( $block['attrs']['isInherited'] ) ) {
+			return $block_content;
+		}
+
+		$block_name = $block['blockName'];
+
+		/**
+		 * The core/post-template has two different block names:
+		 * - core/post-template when the wrapper is rendered.
+		 * - core/null when the loop item is rendered.
+		 */
+		if (
+			'core/null' === $block_name &&
+			isset( $block['attrs']['__woocommerceNamespace'] ) &&
+			'woocommerce/product-query/product-template' === $block['attrs']['__woocommerceNamespace']
+		) {
+			$block_name = self::LOOP_ITEM_ID;
+		}
+
+		$hook_data = $this->get_hook_data();
+
+		if ( ! in_array( $block_name, array_keys( $hook_data ), true ) ) {
+			return $block_content;
+		}
+
+		/**
+		 * If the block is empty, we don't need to inject hooks.
+		 */
+		if ( empty( $block_content ) ) {
+			return $block_content;
+		}
```

Comment:
Will this mean the `woocommerce_no_products_found` won't fire?

Reply:
This makes the actions fire only when blocks return non-empty content. For the `core/query-no-results` case, when the query has products, the block content will be empty, then `woocommerce_no_products_found` shouldn't be fired.

But this will be false positive if we remove all inner blocks from `core/query-no-results`. We need a second thought on this 🤔 .